sources/ec2/ec2_source.go

```diff
@@ -0,0 +1,279 @@
+package ec2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/overmindtech/sdp-go"
+)
+
+const DefaultMaxResultsPerPage = 100
+
+// EC2Souirce This Struct allows us to create sources easily despite the
+// differeneces between the many EC2 APIs. Not that pagniated APIs should
+// populate the `InputMapperPaginated` and `OutputMapperPaginated` fields, where
+// non-paginated APIs should use `InputMapper` and `OutputMapper`. The source
+// will return an error if you use any other combination
+type EC2Source[Input any, Output any] struct {
+	MaxResultsPerPage int32  // Max results per page when making API queries
+	ItemType          string // The type of items that will be returned
+
+	// The funciton that should be used to describe the resources that this
+	// source is related to
+	DescribeFunc func(ctx context.Context, client *ec2.Client, input Input, optFns ...func(*ec2.Options)) (Output, error)
+
+	// A function that returns the input object that will be passed to
+	// DescribeFunc for a given set of scope, query and method
+	InputMapper func(scope, query string, method sdp.RequestMethod) (Input, error)
+
+	// A function that returns the input object will be passed to DescribeFunc
+	// for a given set of scope, query and method. This also includes a
+	// maxResults param which should be fed to the API to mean "page size", and
+	// a nextToken string pointer: If this is the first page this will be nil,
+	// however for subsequest pages this will be populated with the appropriate
+	// token
+	InputMapperPaginated func(scope, query string, method sdp.RequestMethod, maxResults *int32, nextToken *string) (Input, error)
```

Comment:
should this return/wrap [Paginators](https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#using-paginators) where possible? Looking at the paginator implementation, there's more happening than meets the eye, e.g. this `HasMorePages` method has more conditions than our impl does:

```go
func (p *DescribeInstancesPaginator) HasMorePages() bool {
	return p.firstPage || (p.nextToken != nil && len(*p.nextToken) != 0)
}
```